DRILL-8528: HBase Limit Push Down

shfshihuafeng opened the pull request:
# [DRILL-8528](https://issues.apache.org/jira/browse/DRILL-8528): HBase Limit Push Down

## Description

support  Limit Push Down for HBase.i test 

select * from clicks limit 3

The log shows that The storage layer hbase （HBaseRecordReader）only got 3 rows of data

`2025-07-11 01:48:17,297 [178f302e-26db-37dd-500d-db2797b17a5c:frag:0:0] INFO  o.a.d.e.s.hbase.HBaseRecordReader - Took 7 ms to get 3 records`

plan is as follow

```
00-00    Screen : rowType = RecordType(ANY row_key, (VARCHAR(65535), ANY) MAP clickinfo, (VARCHAR(65535), ANY) MAP iteminfo): rowcount = 3.0, cumulative cost = {18.3 rows, 51.3 cpu, 0.0 io, 0.0 network, 0.0 memory}, id = 301
00-01      Project(row_key=[$0], clickinfo=[$1], iteminfo=[$2]) : rowType = RecordType(ANY row_key, (VARCHAR(65535), ANY) MAP clickinfo, (VARCHAR(65535), ANY) MAP iteminfo): rowcount = 3.0, cumulative cost = {18.0 rows, 51.0 cpu, 0.0 io, 0.0 network, 0.0 memory}, id = 300
00-02        SelectionVectorRemover : rowType = RecordType(ANY row_key, (VARCHAR(65535), ANY) MAP clickinfo, (VARCHAR(65535), ANY) MAP iteminfo): rowcount = 3.0, cumulative cost = {15.0 rows, 42.0 cpu, 0.0 io, 0.0 network, 0.0 memory}, id = 297
00-03          Limit(fetch=[3]) : rowType = RecordType(ANY row_key, (VARCHAR(65535), ANY) MAP clickinfo, (VARCHAR(65535), ANY) MAP iteminfo): rowcount = 3.0, cumulative cost = {12.0 rows, 39.0 cpu, 0.0 io, 0.0 network, 0.0 memory}, id = 296
00-04            Scan(table=[[hbase, clicks]], groupscan=[HBaseGroupScan [HBaseScanSpec=HBaseScanSpec [tableName="clicks"], columns=[`row_key`, `clickinfo`, `iteminfo`], maxRecords=3]]) : rowType = RecordType(ANY row_key, (VARCHAR(65535), ANY) MAP clickinfo, (VARCHAR(65535), ANY) MAP iteminfo): rowcount = 9.0, cumulative cost = {9.0 rows, 27.0 cpu, 0.0 io, 0.0 network, 0.0 memory}, id = 295
```


## Documentation
(Please describe user-visible changes similar to what should appear in the Drill documentation.)

## Testing
add test:  org.apache.drill.hbase.TestHBaseFilterPushDown#testLimitPushDown

cgivre commented on contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseGroupScan.java:
```diff
@@ -382,7 +402,7 @@ public String getDigest() {
   public String toString() {
     return "HBaseGroupScan [HBaseScanSpec="
```
Can we update this method with the `PlanStringBuilder` as shown below?   Please update with appropriate fields
```
  @Override
  public String toString() {
    return new PlanStringBuilder(this)
      .field("HbaseScanSpec", hbaseScanSpec)
      .field("columns", columns)
      .field("maxRecords", maxRecords)
      .toString();
  }
```

cgivre commented on contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseGroupScan.java:
```diff
@@ -396,6 +416,33 @@ public List<SchemaPath> getColumns() {
+   * Default is not to support limit pushdown.
+   */
+  @Override
+  @JsonIgnore
```
Is there a reason for the `JsonIgnore`?

cgivre commented on contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseSubScan.java:
```diff
@@ -98,6 +101,11 @@ public HBaseStoragePlugin getStorageEngine(){
   }
 
+  @JsonIgnore
+  public int getMaxRecords() {
```
Again is there a reason for the `JsonIgnore`?

cgivre commented on contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseRecordReader.java:
```diff
@@ -84,12 +84,16 @@ public class HBaseRecordReader extends AbstractRecordReader implements DrillHBas
         Preconditions.checkNotNull(subScanSpec, "HBase reader needs a sub-scan spec").getTableName());
     hbaseScan = new Scan(subScanSpec.getStartRow(), subScanSpec.getStopRow());
     hbaseScanColumnsOnly = new Scan();
+    //todo Whether to remove if condition judgment
```
I'm not sure what this comment means but please remove or explain further.   Just so you are aware, Calcite does provide the logic as to when to push the limit down and when not to so I think we are good here.  To verify this, you can add a unit test with the query:

```sql
SELECT * 
FROM hbase
LIMIT 10 OFFSET 5
```
and you should see the `maxRecords` parameter set to 15. 

If you look at the unit tests for the Phoenix plugin, you can see a good example of this:

https://github.com/apache/drill/blob/78a52087bb85bcc7034157544193155b87503111/contrib/storage-phoenix/src/test/java/org/apache/drill/exec/store/phoenix/PhoenixSQLTest.java#L92-L128

cgivre reviewed (changes requested):
@shfshihuafeng 
Thanks for this.  I have a few minor nits but other than that it looks good.

cgivre commented on contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseRecordReader.java:
```diff
@@ -84,13 +84,15 @@ public class HBaseRecordReader extends AbstractRecordReader implements DrillHBas
     hbaseScanColumnsOnly = new Scan();
+    // Set the limit of rows for this scan. We will terminate the scan if the number of returned rows reaches this value.
     hbaseScan
+        .setLimit(maxRecords)
```
I saw that you removed the logic to test whether the `maxRecords` is defined.   Does this still work if `maxRecords` is `0` or `null`?

cgivre reviewed (approved):
LGTM +1
Thanks for the PR!

shfshihuafeng commented:
> It looks like I spoke too soon. The tests are failing for Java 17.
 the test cases passed In my local environment for java 17,May be  trigger bugs that are difficult to troubleshoot, I will reproduce this issue
<img width="1795" height="537" alt="image" src="https://github.com/user-attachments/assets/488cb7b7-2579-4613-853c-6887ea9f274e" />

shfshihuafeng commented:
> > It looks like I spoke too soon. The tests are failing for Java 17.
> 
> @cgivre This error should not be related to the code (limit push down) submitted this time. I added my unit test cases to the original code,( there is no hbase limit push down function). it still reports the same error



> It looks like I spoke too soon. The tests are failing for Java 17.
@cgivre The error is not caused by submitted code (hbase limit push down. I added my unit test cases to the original code,( there is no hbase limit push down function).  it still reports the same error which occurs occasionally. The plan and error  shows that the limit has not been pushed down. 

<img width="1857" height="804" alt="image" src="https://github.com/user-attachments/assets/140d9412-b6b5-45d6-b152-5c566416185c" />

cgivre commented:
> > It looks like I spoke too soon. The tests are failing for Java 17.
> 
> @cgivre Maybe it would be best to open a new JIRA for this error and add some unit test for offset

I'm going to rerun the failing tests and see if that fixes it.  Sometimes these tests are flaky and will randomly fail.

cgivre reviewed (approved):
LGTM +1.  Thanks for submitting!
